feat: basic support for eta reduction in `grind`

leodemoura opened the pull request:
This PR adds basic support for eta-reduction to `grind`.

leodemoura commented:
cc @kim-em

kim-em commented on tests/lean/run/grind_eta.lean:
```diff
@@ -0,0 +1,15 @@
+  grind [List.foldr_reverse]
+
+/-
+The following example does work yet. `grind` doesn't have a mechanism
```
```suggestion
The following example doesn't work yet. `grind` doesn't have a mechanism
```

kim-em commented:
@leodemoura, I fixed the tests, but could you confirm you're happy with the previous `fail_if_success grind -funext` test now succeeding?

leodemoura commented:
> @leodemoura, I fixed the tests, but could you confirm you're happy with the previous `fail_if_success grind -funext` test now succeeding?

Thanks @kim-em. I got back to this PR, rebased, and put it on the merge queue.
